### PR TITLE
Ability to drag(extrude) an edge

### PR DESCRIPTION
**Pull request by sivako**
Drag from any where between 2 points in edge.
nebula-extrude

### CODE REVIEW

**Comment by supersonicclay** [comment]
Looks pretty good. I think we can make this better by not adding additional points. See this demonstration and compare to how JOSM extrude works:
extrude-bug

**Comment by supersonicclay** [comment]
Also, it would be good i think if clicking an edge adds a point. With JOSM, you double-click to add an intermediate point. But I think it is ok to make it a single click, and should be easier to implement.

**Comment by sivako** [comment]
Thank you @supersonicclay.  great points. I will update accordingly (and adding point). Thanks again.

**Comment by sivako** [comment]
nebula-extrude-2

@supersonicclay, I have updated code addressing comments.  1) while dragging do not add point when 90 corners 2) ability to add point by clicking on edge. (both add and remove point by click will be available for both modify and extrude modes, I think that will be useful for both modes).

File: examples/deck/example.js [hunk 1]
```diff
@@ -329,7 +330,9 @@ export default class Example extends Component<
       // Editing callbacks
       onEdit: ({ updatedData, editType, featureIndex, positionIndexes, position }) => {
         let updatedSelectedFeatureIndexes = this.state.selectedFeatureIndexes;
-        if (!['movePosition', 'rotating', 'translating', 'scaling'].includes(editType)) {
+        if (
+          !['movePosition', 'moveEdge', 'rotating', 'translating', 'scaling'].includes(editType)
```
[the remark this answers is not in the document]
Comment (sivako) [reply examples/deck/example.js]: ok, Thats good idea. I updated to 'extruding' here.

File: modules/core/src/lib/mode-handlers/extrude-handler.js [hunk 2]
```diff
@@ -0,0 +1,234 @@
+// @flow
+
+import bearing from '@turf/bearing';
+import { generatePointsParallelToLinePoints } from '../utils';
+import type { PointerMoveEvent, StartDraggingEvent, StopDraggingEvent } from '../event-types.js';
+import type { EditAction } from './mode-handler.js';
+import { getPickedEditHandle } from './mode-handler.js';
+import { ModifyHandler } from './modify-handler';
+
+export class ExtrudeHandler extends ModifyHandler {
```
[the remark this answers is not in the document]
Comment (sivako) [reply modules/core/src/lib/mode-handlers/extrude-handler.js]: Added tests for ExtrudeHandler and utils.

File: modules/core/src/lib/utils.js [hunk 3]
```diff
@@ -51,3 +57,34 @@ export function recursivelyTraverseNestedArrays(
...
+  p1: Position,
+  p2: Position,
+  groundCoords: Position
+): Position[] {
+  const lineString: LineString = {
+    type: 'LineString',
+    coordinates: [p1, p2]
+  };
+  const pt = point(groundCoords);
+  const options = { units: 'miles' };
```
[the remark this answers is not in the document]
Comment (sivako) [reply modules/core/src/lib/utils.js]: Oh right, No need for options. I have removed.

File: modules/core/src/lib/mode-handlers/extrude-handler.js [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+            editHandle.featureIndex,
+            editHandle.positionIndexes,
+            p1
+          );
+          this.isPointAdded = true;
+        }
+
+        editAction = {
+          updatedData: updatedData.getObject(),
+          editType: 'addEdge',
```
[the remark this answers is not in the document]
Comment (sivako) [reply modules/core/src/lib/mode-handlers/extrude-handler.js]: agree, that make sense. have updated to 'startExtruding', 'extruding' and 'extruded' for start, move and stop events. Updated the doc.

**Comment by sivako** [comment]
@supersonicclay, Fixed the issue with extruding near first/last point. Added the tests and updated with addressing other comments. I have also tested with various scenarios. Can you take a look?, Thank you.

File: docs/api-reference/mode-handlers/overview.md [hunk 5]
```diff
@@ -16,6 +16,12 @@ No edits are possible, but selection is still possible.
 
 User can move existing points, add intermediate points along lines, and remove points.
 
+## [ExtrudeHandler](https://github.com/uber/nebula.gl/blob/master/modules/core/src/lib/mode-handlers/extrude-handler.js)
+
+* Mode name: `extrude`
+
+User can move edge. drag from any where between 2 points in edge.
```
[the remark this answers is not in the document]
Comment (sivako) [reply docs/api-reference/mode-handlers/overview.md]: I just updated.

File: modules/core/src/test/lib/mode-handlers/extrude-handler.test.js [hunk 6]
```diff
@@ -0,0 +1,227 @@
+// @flow
+/* eslint-env jest */
```
[the remark this answers is not in the document]
Comment (sivako) [reply modules/core/src/test/lib/mode-handlers/extrude-handler.test.js]: Agree, I have added the tests for handleClick, handleStartDragging, handlePointerMove as well.

File: docs/api-reference/layers/editable-geojson-layer.md [hunk 7]
```diff
@@ -139,6 +139,12 @@ The `onEdit` event is the core event provided by this layer and must be handled
 
   * `translated`: A feature finished translating (e.g. user finished dragging).
 
+  * `startExtruding`: A feature started extruding.
```
[the remark this answers is not in the document]
Comment (sivako) [reply docs/api-reference/layers/editable-geojson-layer.md]: Ok, sure, I just updated. Thank you.

**Comment by sivako** [comment]
Thank you @supersonicclay @georgios-uber. I have updated doc and tests. Can you please merge to master and create new tag for release?

**Pull request merged**